ImageBufferIOSurfaceBackend should get and put data in WP when in mapped mode

kkinnunen-apple opened the pull request:
#### 812ce301dec978351734e9cd4f2b43a399599747
<pre>
ImageBufferIOSurfaceBackend should get and put data in WP when in mapped mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=252987">https://bugs.webkit.org/show_bug.cgi?id=252987</a>
rdar://problem/105972277

Reviewed by Matt Woodrow.

Make putImageData(), getImageData() access the data directly from the
Web process when IOSurfaces are available there.

Avoids sending messages to GPU process and memcpying the contents
multiple times.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::flushContext):
(WebCore::ImageBuffer::copyNativeImageForDrawing const):
(WebCore::ImageBuffer::draw):
(WebCore::ImageBuffer::getPixelBuffer const):
(WebCore::ImageBuffer::putPixelBuffer):
Before, the ImageBuffer would flush the drawing context,
e.g. the IPC commands to the GPUP and flush the draws there,
and then it would flush the backend context.
Move the backend maintainance into the backend functions, as it is
internal detail of the backend.
Instead just flush the drawing context, which on non-GPUP case
flushes the real IOSurface context and GPUP case flushes through
the IPC.

* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyNativeImageForDrawing):
Provide the destination context to copyNativeImageForDrawing instead
of unused BackingStoreCopy argument.

* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing):
Fix the copyNativeImageForDrawing(). Before, it would return a CGImage
but not mark the m_mayHaveOutstandingBackingStoreReferences. The intention
was that instead, it would call invalidateCachedNativeImage() to cleanup
the bitmap cache from the cached CGImage. However, this would still leave
some backends, like display list and iosurface, to reference the IOSurface
without ImageBufferIOSurfaceBackend knowing about it. In these cases,
putImageData could change the underlying IOSurface before the display list
or iosurface renderer would have completed their use.

Instead, use CGIOSurfaceContextCreateImageReference which has the added
benefit of not needing the fixup case for cleaning up the memory footprint
after bitmap rendering.

(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer):
Flush the writes through the backends drawing context.

(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
Flush the writes through the drawing context and CGImage reads.

(WebCore::ImageBufferIOSurfaceBackend::surface):
(WebCore::ImageBufferIOSurfaceBackend::ensureNativeImagesHaveCopiedBackingStore):
(WebCore::ImageBufferIOSurfaceBackend::prepareForExternalRead):
(WebCore::ImageBufferIOSurfaceBackend::prepareForExternalWrite):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::finalizeDrawIntoContext): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:

* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::copyNativeImageForDrawing const):
(WebKit::RemoteImageBufferProxy::getPixelBuffer const):
(WebKit::RemoteImageBufferProxy::putPixelBuffer):
The remote proxy putPutlBuffer/getPixelBuffer can now directly call the
normal ImageBufferIOSurfaceBackend codepath for the put/get.
The IPC is flushed by the ImageBuffer calling flushDrawingContext()
The pending read references through the local iosurface are maintained
by the ImageBufferIOSurfaceBackend.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
The m_needsFlush needs to be flipped to true, since now the first
command might be local putPixelBuffer. This needs to complete after
the initial CGContextFillRect that clears the IOSurface.
Before this was not needed, as all the commands were executed on GPUP
side.

Canonical link: <a href="https://commits.webkit.org/261130@main">https://commits.webkit.org/261130@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5fbf8ac441b6e563b61740969a5cbaf98c62c705

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19314 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42875 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1636 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119207 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114168 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10505 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102475 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115962 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15468 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98675 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43682 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30339 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85526 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12004 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31676 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12616 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8640 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17982 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51293 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7711 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14425 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
```diff
@@ -79,9 +79,9 @@ class WEBCORE_EXPORT ImageBufferIOSurfaceBackend : public ImageBufferCGBackend {
 
     unsigned bytesPerRow() const override;
 
-    void finalizeDrawIntoContext(GraphicsContext& destinationContext) override;
```
Is this the only impl of this function that's been removed? Can we drop the callsite and decl from ImageBufferBackend.h?

kkinnunen-apple replied:
I left it due to it's not entirely clear yet if we do not need it in the very near future.
We can remove it in a separate patch. I'll do it if nobody else will after near future.

mattwoodrow commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -415,9 +419,15 @@ void ImageBuffer::draw(GraphicsContext& destContext, const FloatRect& destRect,
-        if (auto image = copyNativeImageForDrawing(&destContext == &context() ? CopyBackingStore : DontCopyBackingStore))
-            destContext.drawNativeImage(*image, backendSize(), destRect, srcRectScaled, options);
-        backend->finalizeDrawIntoContext(destContext);
+        if (&destContext == &context()) {
```
Could this logic go inside `copyNativeImageForDrawing`? Especially if we no longer need `finalizeDrawIntoContext`, since the logic is the same either way.

It might be nice for the backend impulse of `copyNativeImageForDrawing` to assert that they never get passed their own context (or alternatively move the logic into the backend, if some backends can handle the same-context case differently).

kkinnunen-apple replied:
I cannot think of a case where the backend could do arbitrary in-place update from itself. That is, a backend that then the draw implementation would be better for the non-copying case than the copying case.

> Could this logic go inside copyNativeImageForDrawing?

It can once RemoteImageBufferProxy is removed and the function is made non-overridable.

Overrides might miss the case and as such have the bug. E.g. today just moving the implementation to the ImageBuffer::copyNativeImageForDrawing would cause the transient bug that RemoteImageBufferProxy::copyNativeImage(DontCopyBackingStore) would be called for GPUP-only image buffers. Today this would not make a difference but tomorrow, who knows!

I'll land as is and we can develop further in further patches.

kkinnunen-apple committed:
ImageBufferIOSurfaceBackend should get and put data in WP when in mapped mode
https://bugs.webkit.org/show_bug.cgi?id=252987
rdar://problem/105972277

Reviewed by Matt Woodrow.

Make putImageData(), getImageData() access the data directly from the
Web process when IOSurfaces are available there.

Avoids sending messages to GPU process and memcpying the contents
multiple times.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::flushContext):
(WebCore::ImageBuffer::copyNativeImageForDrawing const):
(WebCore::ImageBuffer::draw):
(WebCore::ImageBuffer::getPixelBuffer const):
(WebCore::ImageBuffer::putPixelBuffer):
Before, the ImageBuffer would flush the drawing context,
e.g. the IPC commands to the GPUP and flush the draws there,
and then it would flush the backend context.
Move the backend maintainance into the backend functions, as it is
internal detail of the backend.
Instead just flush the drawing context, which on non-GPUP case
flushes the real IOSurface context and GPUP case flushes through
the IPC.

* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::copyNativeImageForDrawing):
Provide the destination context to copyNativeImageForDrawing instead
of unused BackingStoreCopy argument.

* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing):
Fix the copyNativeImageForDrawing(). Before, it would return a CGImage
but not mark the m_mayHaveOutstandingBackingStoreReferences. The intention
was that instead, it would call invalidateCachedNativeImage() to cleanup
the bitmap cache from the cached CGImage. However, this would still leave
some backends, like display list and iosurface, to reference the IOSurface
without ImageBufferIOSurfaceBackend knowing about it. In these cases,
putImageData could change the underlying IOSurface before the display list
or iosurface renderer would have completed their use.

Instead, use CGIOSurfaceContextCreateImageReference which has the added
benefit of not needing the fixup case for cleaning up the memory footprint
after bitmap rendering.

(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer):
Flush the writes through the backends drawing context.

(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
Flush the writes through the drawing context and CGImage reads.

(WebCore::ImageBufferIOSurfaceBackend::surface):
(WebCore::ImageBufferIOSurfaceBackend::ensureNativeImagesHaveCopiedBackingStore):
(WebCore::ImageBufferIOSurfaceBackend::prepareForExternalRead):
(WebCore::ImageBufferIOSurfaceBackend::prepareForExternalWrite):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImageIfNeeded): Deleted.
(WebCore::ImageBufferIOSurfaceBackend::finalizeDrawIntoContext): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:

* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::copyNativeImageForDrawing const):
(WebKit::RemoteImageBufferProxy::getPixelBuffer const):
(WebKit::RemoteImageBufferProxy::putPixelBuffer):
The remote proxy putPutlBuffer/getPixelBuffer can now directly call the
normal ImageBufferIOSurfaceBackend codepath for the put/get.
The IPC is flushed by the ImageBuffer calling flushDrawingContext()
The pending read references through the local iosurface are maintained
by the ImageBufferIOSurfaceBackend.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
The m_needsFlush needs to be flipped to true, since now the first
command might be local putPixelBuffer. This needs to complete after
the initial CGContextFillRect that clears the IOSurface.
Before this was not needed, as all the commands were executed on GPUP
side.

Canonical link: https://commits.webkit.org/261130@main
